# 손글씨 분류하기

In [44]:
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import classification_report

In [45]:
digits = load_digits()
digits_data = digits.data
digits_label = digits.target
digits_data.shape

(1797, 64)

In [46]:
X_train, X_test, y_train, y_test = train_test_split(digits_data, 
                                                    digits_label, 
                                                    test_size=0.2, 
                                                    random_state=13)
X_train.shape, X_test.shape

((1437, 64), (360, 64))

In [47]:
decision_tree = DecisionTreeClassifier(random_state=13)
decision_tree.fit(X_train, y_train)
y_pred = decision_tree.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.97      0.93      0.95        30
           1       0.83      0.78      0.81        37
           2       0.87      0.87      0.87        38
           3       0.79      0.84      0.82        37
           4       0.89      0.87      0.88        38
           5       0.89      0.93      0.91        45
           6       0.95      0.93      0.94        42
           7       0.96      0.92      0.94        26
           8       0.93      0.79      0.86        34
           9       0.72      0.88      0.79        33

    accuracy                           0.88       360
   macro avg       0.88      0.87      0.88       360
weighted avg       0.88      0.88      0.88       360



In [48]:
from sklearn.ensemble import RandomForestClassifier

X_train, X_test, y_train, y_test = train_test_split(digits_data, 
                                                    digits_label, 
                                                    test_size=0.2, 
                                                    random_state=13)

random_forest = RandomForestClassifier(random_state=13)
random_forest.fit(X_train, y_train)
y_pred = random_forest.predict(X_test)

print(classification_report(y_test, y_pred))


              precision    recall  f1-score   support

           0       1.00      1.00      1.00        30
           1       0.95      1.00      0.97        37
           2       1.00      1.00      1.00        38
           3       1.00      0.97      0.99        37
           4       1.00      1.00      1.00        38
           5       1.00      0.98      0.99        45
           6       1.00      1.00      1.00        42
           7       1.00      0.96      0.98        26
           8       1.00      0.91      0.95        34
           9       0.89      1.00      0.94        33

    accuracy                           0.98       360
   macro avg       0.98      0.98      0.98       360
weighted avg       0.98      0.98      0.98       360



In [49]:
from sklearn import svm
svm_model = svm.SVC()

svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        30
           1       0.93      1.00      0.96        37
           2       0.97      1.00      0.99        38
           3       1.00      0.97      0.99        37
           4       1.00      1.00      1.00        38
           5       1.00      0.98      0.99        45
           6       1.00      1.00      1.00        42
           7       1.00      0.96      0.98        26
           8       1.00      0.91      0.95        34
           9       0.94      1.00      0.97        33

    accuracy                           0.98       360
   macro avg       0.98      0.98      0.98       360
weighted avg       0.98      0.98      0.98       360



In [50]:
from sklearn.linear_model import SGDClassifier
sgd_model = SGDClassifier()

sgd_model.fit(X_train, y_train)
y_pred =sgd_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        30
           1       0.74      1.00      0.85        37
           2       0.97      0.97      0.97        38
           3       0.97      0.92      0.94        37
           4       0.97      1.00      0.99        38
           5       1.00      0.96      0.98        45
           6       0.98      0.98      0.98        42
           7       0.93      0.96      0.94        26
           8       1.00      0.59      0.74        34
           9       0.89      0.97      0.93        33

    accuracy                           0.94       360
   macro avg       0.95      0.93      0.93       360
weighted avg       0.95      0.94      0.93       360



In [51]:
from sklearn.linear_model import LogisticRegression
logistic_model = LogisticRegression(max_iter=5000)

logistic_model.fit(X_train, y_train)
y_pred = logistic_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        30
           1       0.85      0.95      0.90        37
           2       1.00      0.95      0.97        38
           3       0.97      0.92      0.94        37
           4       0.97      1.00      0.99        38
           5       0.98      0.96      0.97        45
           6       1.00      0.98      0.99        42
           7       0.96      0.96      0.96        26
           8       0.97      0.82      0.89        34
           9       0.85      1.00      0.92        33

    accuracy                           0.95       360
   macro avg       0.95      0.95      0.95       360
weighted avg       0.96      0.95      0.95       360



In [52]:
from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test, y_pred)
accuracy

0.9527777777777777

0.9527777777777777의 정확도를 보입니다.


accuracy(정확도) 평가지표를 사용한 이유:

손글씨를 정확하게 분류했는지 확인하기 위해서 입니다.

# 와인 분류하기

In [53]:
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import classification_report

In [54]:
wine = load_wine()
wine_data = wine.data
wine_label = wine.target
wine_data.shape

(178, 13)

In [55]:
X_train, X_test, y_train, y_test = train_test_split(wine_data, 
                                                    wine_label, 
                                                    test_size=0.2, 
                                                    random_state=13)
X_train.shape, X_test.shape

((142, 13), (36, 13))

In [56]:
decision_tree = DecisionTreeClassifier(random_state=13)
decision_tree.fit(X_train, y_train)
y_pred = decision_tree.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.86      1.00      0.92        12
           1       0.87      0.87      0.87        15
           2       1.00      0.78      0.88         9

    accuracy                           0.89        36
   macro avg       0.91      0.88      0.89        36
weighted avg       0.90      0.89      0.89        36



In [57]:
from sklearn.ensemble import RandomForestClassifier

X_train, X_test, y_train, y_test = train_test_split(wine_data, 
                                                    wine_label, 
                                                    test_size=0.2, 
                                                    random_state=13)

random_forest = RandomForestClassifier(random_state=13)
random_forest.fit(X_train, y_train)
y_pred = random_forest.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        12
           1       1.00      1.00      1.00        15
           2       1.00      1.00      1.00         9

    accuracy                           1.00        36
   macro avg       1.00      1.00      1.00        36
weighted avg       1.00      1.00      1.00        36



In [58]:
from sklearn import svm
svm_model = svm.SVC()

svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        12
           1       0.73      0.73      0.73        15
           2       0.56      0.56      0.56         9

    accuracy                           0.78        36
   macro avg       0.76      0.76      0.76        36
weighted avg       0.78      0.78      0.78        36



In [75]:
from sklearn.linear_model import SGDClassifier
sgd_model = SGDClassifier()

sgd_model.fit(X_train, y_train)
y_pred =sgd_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.67      0.80        36
           1       0.87      1.00      0.93        78

    accuracy                           0.89       114
   macro avg       0.93      0.83      0.86       114
weighted avg       0.91      0.89      0.89       114



In [60]:
from sklearn.linear_model import LogisticRegression
logistic_model = LogisticRegression(max_iter=5000)

logistic_model.fit(X_train, y_train)
y_pred = logistic_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        12
           1       0.94      1.00      0.97        15
           2       1.00      0.89      0.94         9

    accuracy                           0.97        36
   macro avg       0.98      0.96      0.97        36
weighted avg       0.97      0.97      0.97        36



In [61]:
from sklearn.metrics import confusion_matrix

confusion_matrix = confusion_matrix(y_test, y_pred)
confusion_matrix

array([[12,  0,  0],
       [ 0, 15,  0],
       [ 0,  1,  8]])

In [62]:
from sklearn.metrics import multilabel_confusion_matrix

multilabel_confusion_matrix = multilabel_confusion_matrix(y_test, y_pred)

multilabel_confusion_matrix

array([[[24,  0],
        [ 0, 12]],

       [[20,  1],
        [ 0, 15]],

       [[27,  0],
        [ 1,  8]]])

In [63]:
from sklearn.metrics import accuracy_score

accuracy_score = accuracy_score(y_test, y_pred)
accuracy_score

0.9722222222222222

0.9722222222222222의 정확도를 보입니다.

accuracy 평가지표를 쓴 이유:

와인 분류의 정확도를 확인해야 하기 때문에 accuracy 평가 지표를 사용하였습니다.

In [64]:
from sklearn.metrics import balanced_accuracy_score

balanced_accuracy_score = balanced_accuracy_score(y_test, y_pred)
balanced_accuracy_score

0.9629629629629629

# 유방암 판별하기

In [65]:
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import classification_report

In [66]:
breast_cancer = load_breast_cancer()
breast_cancer_data = breast_cancer.data
breast_cancer_label = breast_cancer.target
breast_cancer_data.shape

(569, 30)

In [67]:
X_train, X_test, y_train, y_test = train_test_split(breast_cancer_data, 
                                                    breast_cancer_label, 
                                                    test_size=0.2, 
                                                    random_state=13)
X_train.shape, X_test.shape

((455, 30), (114, 30))

In [68]:
decision_tree = DecisionTreeClassifier(random_state=13)
decision_tree.fit(X_train, y_train)
y_pred = decision_tree.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.77      0.92      0.84        36
           1       0.96      0.87      0.91        78

    accuracy                           0.89       114
   macro avg       0.86      0.89      0.87       114
weighted avg       0.90      0.89      0.89       114



In [69]:
from sklearn.ensemble import RandomForestClassifier

X_train, X_test, y_train, y_test = train_test_split(breast_cancer_data, 
                                                    breast_cancer_label, 
                                                    test_size=0.2, 
                                                    random_state=13)

random_forest = RandomForestClassifier(random_state=13)
random_forest.fit(X_train, y_train)
y_pred = random_forest.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.82      0.92      0.87        36
           1       0.96      0.91      0.93        78

    accuracy                           0.91       114
   macro avg       0.89      0.91      0.90       114
weighted avg       0.92      0.91      0.91       114



In [70]:
from sklearn import svm
svm_model = svm.SVC()

svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.84      0.75      0.79        36
           1       0.89      0.94      0.91        78

    accuracy                           0.88       114
   macro avg       0.87      0.84      0.85       114
weighted avg       0.88      0.88      0.88       114



In [71]:
from sklearn.linear_model import SGDClassifier
sgd_model = SGDClassifier()

sgd_model.fit(X_train, y_train)
y_pred =sgd_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.67      0.80        36
           1       0.87      1.00      0.93        78

    accuracy                           0.89       114
   macro avg       0.93      0.83      0.86       114
weighted avg       0.91      0.89      0.89       114



In [72]:
from sklearn.linear_model import LogisticRegression
logistic_model = LogisticRegression(max_iter=5000)

logistic_model.fit(X_train, y_train)
y_pred = logistic_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.89      0.86      0.87        36
           1       0.94      0.95      0.94        78

    accuracy                           0.92       114
   macro avg       0.91      0.90      0.91       114
weighted avg       0.92      0.92      0.92       114



In [73]:
from sklearn.metrics import confusion_matrix

confusion_matrix = confusion_matrix(y_test, y_pred)
confusion_matrix

array([[31,  5],
       [ 4, 74]])

In [74]:
from sklearn.metrics import recall_score

recall_score = recall_score(y_test, y_pred)

recall_score

0.9487179487179487

재현율은 0.9487179487179487 입니다.


Recall 평가지표를 사용한 이유:

Recall(재현율)은 실제 값이 True인 것들 중 모델이 True로 예측한 비율입니다. 
암예측의 경우 양성을 놓지면 안 되기 때문에 Recall을 사용합니다.

FN의 경우가 적어야 하기 때문에 Recall 평가지표를 사용하여 TP을 높이고 FN을 낮추는 데 초점을 맞출 수 있도록 하였습니다.

느낀점

분류성분 평가지표에 대해 공부할 수 있는 챕터였습니다.
생소하기도 했고 헷갈리는 개념이라 처음 봤을때는 많이 어려웠는데
인터넷 검색과 유튭 동영상을 통해 좀 더 쉽게 공부할 수 있어습니다.

특히 Recall(재현성)은 양성을 놓치면 안될 때 사용한다는 것과 Precision(정밀도)은 음성을 놓치면 안될 때 사용한다는 것을 이해하는데 시간을 많이 썼지만 어쨌든 이해했습니다.
평가지표를 통해 분류 결정 임계값, 즉 Threshold를 조정할 수 있다는 점도 알게 되었습니다.
평가지표를 구하는 이유는 결국 Threshold를 조절해서 더 나은 모델을 만드는 것이라는 걸 알게되니 왜 분류평가를 하는지 이해가 됐습니당.

Scikit learn 에서 제공하는 분류 평가 지표가 많아서 다 해보고싶었는데 자료별로 사용할 수 있는 것도 있고 안 되는 것도 있다는 것을 알게 되었습니다

재미있었습니다.ㅎ